examples/gosherve-demo/src/charm.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+# Copyright 2021 Jon Seager
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/
+
+"""Charm the service.
+
+Refer to the following tutorial for a quick-start guide that will help you
+develop a new K8s charm using Ops:
+
+    https://ops.readthedocs.io/en/latest/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/index.html
+"""
+
+import dataclasses
+import logging
+import urllib.request
+
+import ops
+from charms.traefik_k8s.v1.ingress import IngressPerAppRequirer
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class GosherveConfig:
+    """Configuration for the Gosherve charm."""
+
+    redirect_map: str = (
+        'https://github.com/canonical/operator/tree/main/examples/gosherve-demo/demo-routes'
+    )
+    """A URL pointing to a list of redirects for Gosherve."""
+
+    def __post_init__(self):
+        """Validate the configuration."""
+        if not self.redirect_map:
+            raise ValueError('redirect-map must be set in the configuration.')
+        if not self.redirect_map.startswith('http'):
+            raise ValueError(
+                f'Invalid redirect-map URL: {self.redirect_map}. '
+                "It must start with 'http' or 'https'."
+            )
+
+
+class HelloKubeconCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        framework.observe(self.on.install, self._on_install)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on['gosherve'].pebble_ready, self._on_config_changed)
+        framework.observe(self.on['pull_site'].action, self._pull_site_action)
+        self.ingress = IngressPerAppRequirer(
+            self,
+            port=8080,
+            host=f'{self.app.name}-endpoints.{self.model.name}.svc.cluster.local',
+            strip_prefix=True,
+        )
+
+    def _on_install(self, _: ops.InstallEvent):
+        # Download the site:
+        self._fetch_site()
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent | ops.PebbleReadyEvent):
```

Comment:
Would it be better to call this:

```suggestion
    def _reconcile(self, _: ops.ConfigChangedEvent | ops.PebbleReadyEvent):
```